File: Hackontrol/libnative32/Kernel.c
#include "Kernel.h"

static JNINativeMethod NativeMethods[] = {
	{"sleep",                       "()V",                                                 (void*) &Kernel_sleep},
	{"hibernate",                   "()V",                                                 (void*) &Kernel_hibernate},
	{"restart",                     "()V",                                                 (void*) &Kernel_restart},
	{"shutdown",                    "()V",                                                 (void*) &Kernel_shutdown},
	{"setProcessCritical",          "(Z)V",                                                (void*) &Kernel_setProcessCritical},
	{"setFreeze",                   "(Z[B)V",                                              (void*) &Kernel_setFreeze},
	{"getProcessList",              "()[Lcom/khopan/hackontrol/ProcessEntry;",             (void*) &Kernel_getProcessList},
	{"getCurrentProcessIdentifier", "()I",                                                 (void*) &Kernel_getCurrentProcessIdentifier},
	{"terminateProcess",            "(I)V",                                                (void*) &Kernel_terminateProcess},
	{"initiateRestart",             "(Z)V",                                                (void*) &Kernel_initiateRestart},
	{"shellExecute",                "(Ljava/lang/String;Ljava/util/function/Consumer;)V",  (void*) &Kernel_shellExecute},
	{"connectHRSPServer",           "(Ljava/lang/String;ILjava/util/function/Consumer;)V", (void*) &Kernel_connectHRSPServer}
};

void KernelRegisterNatives(JNIEnv* const environment) {
	jclass class = (*environment)->FindClass(environment, "com/khopan/hackontrol/library/Kernel");

	if(!class) {
		return;
	}

	(*environment)->RegisterNatives(environment, class, NativeMethods, sizeof(NativeMethods) / sizeof(NativeMethods[0]));
}

File: src/com/khopan/hackontrol/library/Kernel.java
package com.khopan.hackontrol.library;

import java.util.function.Consumer;

import com.khopan.hackontrol.ProcessEntry;

public class Kernel {
	private Kernel() {}

	public static native void sleep();
	public static native void hibernate();
	public static native void restart();
	public static native void shutdown();
	public static native void setProcessCritical(boolean critical);
	public static native void setFreeze(boolean freeze, byte[] image);
	public static native ProcessEntry[] getProcessList();
	public static native int getCurrentProcessIdentifier();
	public static native void terminateProcess(int processIdentifier);
	public static native void initiateRestart(boolean update);
	public static native void shellExecute(String command, Consumer<String> callback);
	public static native void connectHRSPServer(String host, int port, Consumer<String> callback);
}
